### ISSUE TITLE
WarpX nvcc linking stalls on Juwels

### ISSUE THREAD

**Issue by MaxThevenet** [ECP-WarpX/WarpX#1443]
When compiling WarpX on Juwels at JSC, the compilation works fine but the nvcc linking hangs forever (step `Linking main3d.gnu.TPROF.MTMPI.CUDA.ex ...` never finishes).
My `warp.profile` looks like (I removed things I think don't matter)
```
#!/usr/bin/env sh
export SCRATCH=$SCRATCH_chhh45/Maxence
alias sqs="squeue -u $USERNAME"
shopt -s direxpand
module load GCC
module load OpenMPI
module load CUDA
export GPUS_PER_SOCKET=2
export GPUS_PER_NODE=4
```
`module list` returns
```
  1) GCCcore/.8.3.0 (H)   4) GCC/8.3.0            7) XZ/.5.2.4      (H)  10) expat/.2.2.6     (H)  13) gperf/.3.1         (H)  16) fontconfig/.2.13.1 (H)  19) UCX/1.7.0            22) nvidia/.driver (H,g)
  2) binutils/.2.32 (H)   5) zlib/.1.2.11   (H)   8) libxml2/.2.9.9 (H)  11) libpng/.1.6.36   (H)  14) ncurses/.6.1       (H)  17) X11/20190311            20) libevent/.2.1.8 (H)  23) CUDA/10.1.105  (g)
  3) StdEnv/2019a         6) numactl/2.0.12       9) bzip2/.1.0.6   (H)  12) freetype/.2.10.0 (H)  15) util-linux/.2.33.1 (H)  18) hwloc/2.1.0             21) OpenMPI/4.0.2
```
A few additional points:
- WarpX compiles fine on other platforms, of course
- When compiling with CMake, the same problem happens, but at least it seems that the linking of AMReX works successfully (we get message `[ 54%] Linking CUDA static library ../../../lib/libamrex.a`)
- Hipace compilation and linking work well, with the same options, so the problem really seems to be in WarpX.

I will ask if some other versions of gcc and nvcc are available, but it doesn't seem so.

**Comment by ax3l** [comment]
Thank you for the details and JSC support mail kick-off.

For completion, can you please post the last lines shown by `make VERBOSE=ON` and also the full output of `cmake [...]` when run in a fresh/empty build directory?

Things we already tried (on Slack):
- [x] serial build
- [x] CUDA 10.1 supports GCC 8: https://gist.github.com/ax3l/9489132
- [ ] build in tmpfs
- [ ] explicitly set CUDA SM

**Comment by MaxThevenet** [comment]
So after further tests, the code compiles successfully on CUDA/11 and GCC/9.3, and the linking doesn't hang anymore. Unfortunately, these are experimental modules on Juwels, so we cannot run a simulation with this executable. Unsurprisingly, the code compiles fine on CPU.

I know that the code worked fine at the beginning of the month on commit https://github.com/ECP-WarpX/WarpX/commit/73caf7421f9a095468847ff7ce0b8181b09df75c.

Here's the [CMake output](https://github.com/ECP-WarpX/WarpX/files/5391267/CMake_output.txt).

Compiling with `make VERBOSE=ON` doesn't give more information than just `make`. I could try CUDA_VERBOSE=TRUE.

**Comment by MaxThevenet** [comment]
Lewis (@lboult) went through the process of `git-bisect`ing the issue (thanks a lot for that!), and found that the commit that introduced the issue https://github.com/ECP-WarpX/WarpX/commit/02978bc8e4ecc12a74bceeb38e5dbc82615c75e2,  from PR https://github.com/ECP-WarpX/WarpX/pull/1156.

@WeiqunZhang @ax3l do you see what could have cause the linking error (with GCC/8.3.0 and CUDA/10.1.105) from this commit? Lewis can help do some more tests, as he uses WarpX on Juwels for production runs.

**Comment by lboult** [comment]
Hey guys, yes please let me know if I can be of more help with tests!

**Comment by MaxThevenet** [comment]
Update: I hacked a little bit into `Source/Parser/wp_parser_c.h`, and producing a dummy code where `wp_ast_eval` take only one template parameter `template <int Depth>` instead of 2 including 1 optional `template <int Depth, std::enable_if_t<(Depth<WARPX_PARSER_DEPTH), int> = 0>` fixed the issue. However, I had to do other stupid things for this to work, so I don't know precisely what fixed the issue, but with [these changes](https://github.com/ECP-WarpX/WarpX/compare/development...MaxThevenet:test_parser) the code compiles and links properly (although this obviously breaks the parser). @WeiqunZhang @ax3l does that help? Do you see a fix? These optional template parameters are a bit technical for me, so I did not manage to propose an actual fix here.

**Comment by AndiH** [comment]
Jumping in here from JSC!

> Unfortunately, these are experimental modules on Juwels.
That's not true. Although not part of the production stage of software, we will make it the production stage in a few weeks. Right now, `Stages/2019a` is the default, in a few weeks, `Stages/2020` is the default which you can start using already now. It's 80% complete.

No changes were made on the system with software from `Stages/2019a`. We try to leave production stages alone. I think it's more likely that something in your code changed…

The commits you posted are quite long. As an outsider it's hard to judge what could be the culprit. But I think you already made progress? My idea would have been to find out if `AMREX_DEVICE_COMPILE` is always defined correctly.

**Comment by MaxThevenet** [comment]
Thanks @AndiH for coming to help. I looked at the code but couldn't find an obvious fix that would still preserve the code capabilities. In the end, I created a new branch from an old working WarpX commit, and cherry-picked recent changes needed by @lboult. This is available on that branch https://github.com/MaxThevenet/WarpX/tree/lewis. @lboult confirmed that it fixes the issue.

This should be sufficient until `Stages/2020` becomes the default in a few weeks, which will fix our problem. Hence, we can consider this issue closed, and just remember that we should avoid compiling WarpX with `GCC/8.3.0` and `CUDA/10.1.105`.

**Comment by AndiH** [comment]
Alright! I'll close the issue on our side as well.

You can always already use the 2020-Stages with
```
module use $OTHERSTAGES
module load Stages/2020
```
… but I think you know that already.

**Comment by MaxThevenet** [comment]
@AndiH actually, could you keep it open, at least this morning? We were fooled by the test we did, it worked for thousands of iterations but then crashed shamelessly... @lboult will give us an update this morning.

As for
```
module use $OTHERSTAGES
module load Stages/2020
```
I tried that, and the code compiles fine, but it crashes at runtime. We concluded that this was still an experimental installation, not yet ready for production, so we should just wait a bit before using `Stages/2020`. Is that OK with you? Or do you want us to investigate this crash, and document this in the open JSC ticket?

**Comment by AndiH** [comment]
Ah, too late – but I can reopen it; no problem.

The new stage should be about ready to go, so any crash you get right now is not expected. We've just not _released_ the stage yet because we have not reached the end of the list of software we need to provide.
Please report your crash to the ticket you filed (actually, that will reopen it). Any analysis you can already do on your part is greatly appreciated!

**Comment by MaxThevenet** [comment]
Ok I'll re-create the bug, document it and provide a reproducer.

**Comment by MaxThevenet** [comment]
So after more cherry-picking, Lewis is able to run his simulations on Juwels. Thanks for your help!

**Comment by AndiH** [comment]
Alright! That does mean I can close the ticket again, right?

**Comment by MaxThevenet** [comment]
Wait, now I am confused:

**Problem 1**: Linking hangs forever, so we can't do our production runs. Going back to an old working version and cherry-picking required commits provided a sufficient workaround for now, so: **problem solved**

**Problem 2**: with `Stages/2020`, WarpX compiles but crashes at runtime: **problem NOT solved**

You proposed:
> Please report your crash to the ticket you filed (actually, that will reopen it). Any analysis you can already do on your part is greatly appreciated!

so I did that as a continuation of the ticket. Didn't you receive my messages? So unless I misunderstood, you proposed to keep the ticket open until problem 2 is fixed, right?

Regardless, we can close this Github issue, as it is about problem 1, which has a satisfying temporary fix.

**Comment by AndiH** [comment]
You said in your previous comment
> So after more cherry-picking, Lewis is able to run his simulations on Juwels. Thanks for your help!

I assumed that was referring to "**Problem 2**". If not, what was it referring to?

And yes, I was referring to the details you send around 9:35 today, reopening ticket 1041033.

**Comment by MaxThevenet** [comment]
Alright, sorry, that was still problem 1. It was an update to my comment this morning:
> @AndiH actually, could you keep it open, at least this morning? We were fooled by the test we did, it worked for thousands of iterations but then crashed shamelessly... @lboult will give us an update this morning.

so problem 2 is still not fixed, and any help to fix it would still be much appreciated.

**Comment by AndiH** [comment]
Ok! Got it!

**Comment by ax3l** [comment]
I realized a few weeks later: we could-have worked around this by just using another linker with CMake.

By default, CMake uses for CUDA projects the host compiler's toolchain as linker, which is usually the best choice. We saw that our GNUmake build system did not hang, as it linked with nvcc. The corresponding CMake option to change the default linker is:
```bash
cmake -DCMAKE_LINKER=/path/to/linker -DCMAKE_CXX_LINK_EXECUTABLE="<CMAKE_LINKER> <FLAGS> <CMAKE_CXX_LINK_FLAGS> <LINK_FLAGS> <OBJECTS> -o <TARGET> <LINK_LIBRARIES>"
```

So likely `-DCMAKE_LINKER=$(which nvcc)` would have worked as a temporary work-around. Just documenting this here as a good open-source citizen for future readers.

https://stackoverflow.com/a/25274328/2719194
https://github.com/Kitware/CMake/blob/v3.19.1/Modules/CMakeCXXInformation.cmake#L273-L276
